packages/babel-generator/src/buffer.ts

```diff
@@ -27,23 +36,78 @@ function SourcePos(): SourcePos {
   };
 }
 
-const SPACES_RE = /^[ \t]+$/;
 export default class Buffer {
   constructor(map?: SourceMap | null) {
     this._map = map;
+
+    this._allocQueue();
   }
 
   _map: SourceMap = null;
   _buf = "";
+  _str = "";
+  _appendCount = 0;
   _last = 0;
-  _queue: Parameters<Buffer["_append"]>[] = [];
+  _queue: queueObj[] = [];
+  _queueCursor = 0;
 
   _position = {
     line: 1,
     column: 0,
   };
   _sourcePosition = SourcePos();
-  _disallowedPop: SourcePos | null = null;
+  _disallowedPop: SourcePos & { ok: boolean } = {
+    identifierName: undefined,
+    line: undefined,
+    column: undefined,
+    filename: undefined,
+    ok: false,
+  };
+
+  _allocQueue() {
+    const queue = this._queue;
+
+    for (let i = 0; i < 16; i++) {
```

Comment:
Why is it faster to allocate 16 chunks, even if we are still allocating them with `.push` one by one?

Reply:
We rarely call it, and in most cases directly write the value of the array element object.

https://github.com/babel/babel/blob/c944264a89a3678623d991f74b2250e6a4bc4688/packages/babel-generator/src/buffer.ts#L90-L102